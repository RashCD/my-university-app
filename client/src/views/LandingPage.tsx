import { RouteComponentProps } from '@reach/router';
import AppHeader from '../components/AppHeader';
import Input from '../components/Input';
import { useForm } from 'react-hook-form';

import Styles from '../assets/styles/views/LandingPage.module.scss';
import CTAButton from '../components/CTAButton';

type FormValues = {
  search: string;
};

const LandingPage = (props: RouteComponentProps) => {
  const { register, handleSubmit } = useForm<FormValues>();

  const onFormSubmit = (data: FormValues) => {
    console.log(data);
  };

  return (
    <div className="app">
      <AppHeader title="University List App" />
      <main className={Styles.appMain}>
        <form className={Styles.form} onSubmit={handleSubmit(onFormSubmit)}>
          <Input
            id="uni-name"
            label="Name"
            placeholder="Search by University Name"
            register={register('search')}
          />
          <Input
            id="uni-country"
            label="Country"
            placeholder="Search by University Country"
            register={register('search')}
          />
          <CTAButton type="submit">Search</CTAButton>
        </form>
      </main>
    </div>
  );
};

export default LandingPage;
