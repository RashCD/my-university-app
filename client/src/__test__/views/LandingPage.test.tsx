import React from 'react';
import {
  cleanup,
  findByText,
  fireEvent,
  getAllByRole,
  render,
  screen,
  act,
} from '@testing-library/react';
import LandingPage from '../../views/LandingPage';
import {
  createHistory,
  createMemorySource,
  LocationProvider,
} from '@reach/router';

afterEach(cleanup);

function renderWithRouter(
  ui: JSX.Element,
  { route = '/', history = createHistory(createMemorySource(route)) } = {}
) {
  return {
    ...render(<LocationProvider history={history}>{ui}</LocationProvider>),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  };
}

jest.mock('react-query', () => ({
  useQuery: () => ({ isLoading: false, error: {}, data: [] }),
}));

describe('Landing page component test', () => {
  beforeEach(() => {
    renderWithRouter(<LandingPage />);
  });

  test('should contains "University List App"', async () => {
    expect(
      await findByText(document.body, 'University List App')
    ).toHaveTextContent('University List App');
  });

  test('should have Login icon visible', async () => {
    const container = document.body;

    const loginButton = getAllByRole(container, 'button').find(
      (element) => element.className === 'button loginButton'
    );

    expect(loginButton).toBeValid();

    expect(loginButton).toBeVisible();
  });

  test('should go to login page when clicking the login button', () => {
    const container = document.body;

    const loginButton = getAllByRole(container, 'button').find(
      (element) => element.className === 'button loginButton'
    ) as HTMLElement;

    fireEvent.click(loginButton);

    expect(window.location.pathname).toBe('/login');
  });

  test('should show table when submitting the form', async () => {
    const nameField = screen.getByLabelText('Name') as HTMLInputElement;

    const countryField = screen.getByLabelText('Country') as HTMLInputElement;

    const buttonField = screen.getByText('Search') as HTMLButtonElement;

    fireEvent.change(nameField, { target: { value: 'name test' } });

    fireEvent.change(countryField, { target: { value: 'country test' } });

    await act(async () => {
      fireEvent.submit(buttonField);
    });

    // console.log(document.body.innerHTML);

    // await expect(screen.getByRole('table')).toBeInTheDocument();

    // expect(nameField.value).toBe('name test');

    // expect(countryField.value).toBe('country test');
  });

  test('should match the snapshot', () => {
    const container = document.body;

    expect(container).toMatchSnapshot();
  });
});
