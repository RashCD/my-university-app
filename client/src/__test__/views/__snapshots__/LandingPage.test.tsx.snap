// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Landing page component test should match the snapshot 1`] = `
<body>
  <div>
    <div
      class="app"
    >
      <header
        class="appHeader"
      >
        <button
          class="button backButton"
          type="button"
        >
          <span
            class="iconContainer"
            data-icon="true"
            style="--icon-src: url(back.svg); --icon-color: var(--white); --icon-size: 30px;"
          />
          BACK
        </button>
        <h1
          class="title"
        >
          University List App
        </h1>
      </header>
      <main
        class="app-main"
      >
        <div
          class="container"
        >
          <form
            class="form"
          >
            <div
              class="inputContainer"
            >
              <label
                class="label"
                for="uni-name"
              >
                Name
              </label>
              <input
                class="input"
                id="uni-name"
                name="name"
                placeholder="Search by University Name"
                type="text"
              />
            </div>
            <div
              class="inputContainer"
            >
              <label
                class="label"
                for="uni-country"
              >
                Country
              </label>
              <input
                class="input"
                id="uni-country"
                name="country"
                placeholder="Search by University Country"
                type="text"
              />
            </div>
            <button
              class="button cta"
              type="submit"
            >
              Search
            </button>
          </form>
        </div>
      </main>
    </div>
  </div>
</body>
`;
